timeutil/schedule.go

```diff
@@ -215,3 +395,341 @@ func ParseSchedule(scheduleSpec string) ([]*Schedule, error) {
+	count uint
+	// should event time be randomized
+	randomize bool
+}
+
+// makeTimeSchedule creates a single
+func makeTimeSchedule(conf timeScheduleConf, base *Schedule) *Schedule {
+	if base == nil {
+		base = &Schedule{}
+	}
+	nsch := *base
+	nsch.Start = conf.start
+	nsch.End = conf.end
+	nsch.Randomize = conf.randomize
+	return &nsch
+}
+
+// parseWeekdaySpec will parse a string and extract weekday (eg. wed),
+// MonthWeekday (eg. 1st, 5th in the month).
+func parseWeekdaySpec(s string) (string, MonthWeekday, error) {
+	l := len(s)
+	if l != 3 && l != 4 {
+		return "", monthWeekdayNone, fmt.Errorf("cannot parse %q: invalid format", s)
+	}
+
+	wday := s
+	mw := monthWeekdayNone
+	if l == 4 {
+		wday = s[0:3]
+		if v, err := strconv.ParseUint(s[3:], 10, 32); err != nil {
+			return "", monthWeekdayNone, fmt.Errorf("cannot parse %q: invalid week", s)
+		} else {
+			mw = MonthWeekday(v)
+		}
+		if mw < monthWeekdayMin || mw > monthWeekdayMax {
+			return "", monthWeekdayNone, fmt.Errorf("cannot parse %q: incorrect week number", mw)
+		}
+	}
+	if !IsValidWeekday(wday) {
+		return "", monthWeekdayNone, fmt.Errorf("cannot parse %q: invalid weekday", s)
+	}
+
+	return wday, mw, nil
+}
+
+// makeTimeSchedules returns a slice of Schedules generated from the template
+// list. The schedules are filled with time information passed in `conf`. Note,
+// that there may be more schedules generated than passed as a template.
+func makeTimeSchedules(template []*Schedule, conf timeScheduleConf) []*Schedule {
+
+	scheds := []*Schedule{}
+	for _, tsch := range template {
+		// Cut up the 'count' schedules (eg. 9:00-11:00/2 - twice
+		// between 9 and 11) into separate schedules. This way we don't
+		// have to keep any state data to be able to tell how many times
+		// the event has already happened and each event becomes separate.
+		if conf.count > 1 {
+			step := time.Duration(uint64(conf.end.Sub(conf.start)) / uint64(conf.count))
+			tempConf := conf
+			tempConf.start = conf.start
+			for i := uint(0); i < conf.count; i++ {
+				tempConf.end = tempConf.start.Add(step)
+				scheds = append(scheds,
+					makeTimeSchedule(tempConf, tsch))
+				tempConf.start = tempConf.end
+			}
+		} else {
+			scheds = append(scheds,
+				makeTimeSchedule(conf, tsch))
+		}
+	}
+	return scheds
+}
+
+// isValidWeekdaySpec returns true if string is of the following format:
+//     wday =  ( "sun" / "mon" / "tue" / "wed" / "thu" / "fri" / "sat" ) [ DIGIT ]
+func isValidWeekdaySpec(s string) bool {
+	_, _, err := parseWeekdaySpec(s)
+	return err == nil
+}
+
+const (
+	spanToken           = "-"
+	randomizedSpanToken = "~"
+	countToken          = "/"
+)
+
+var specialTokens = map[string]string{
+	"day":     "0:00-24:00",
+	"weekend": "sat-sun",
+	// shorthand variant of whole day
+	"-": "0:00-24:00",
+	// and randomized whole day
+	"~": "0:00~24:00",
+}
+
+// Parse each event and return a slice of schedules.
+func parseEventSetScheduleV2(s string) ([]*Schedule, error) {
```

Comment:
This function got slightly big.. Any chance to move some of the logic in if/else blocks into helper functions?